solvers/SimpleSOL/Diagnostics/momentum_conservation.hpp

```diff
@@ -0,0 +1,307 @@
+    auto particle_group = this->particle_sys->particle_group;
+    auto k_ion_p0 = (*particle_group)[Sym<REAL>("ION_MOMENTUM_0")]
+                        ->cell_dat.device_ptr();
+
+    this->dh_particle_total_momentum_0.h_buffer.ptr[0] = 0.0;
+    this->dh_particle_total_momentum_0.host_to_device();
+    auto k_particle_momentum_0 = this->dh_particle_total_momentum_0.d_buffer.ptr;
+
+    const auto pl_iter_range =
+        particle_group->mpi_rank_dat->get_particle_loop_iter_range();
+    const auto pl_stride =
+        particle_group->mpi_rank_dat->get_particle_loop_cell_stride();
+    const auto pl_npart_cell =
+        particle_group->mpi_rank_dat->get_particle_loop_npart_cell();
+
+    sycl_target->queue
+        .submit([&](sycl::handler &cgh) {
+          cgh.parallel_for<>(
+              sycl::range<1>(pl_iter_range), [=](sycl::id<1> idx) {
+                NESO_PARTICLES_KERNEL_START
+                const INT cellx = NESO_PARTICLES_KERNEL_CELL;
+                const INT layerx = NESO_PARTICLES_KERNEL_LAYER;
+
+                const double contrib = k_ion_p0[cellx][0][layerx];
+
+                sycl::atomic_ref<double, sycl::memory_order::relaxed,
+                                 sycl::memory_scope::device>
+                    energy_atomic(k_particle_momentum_0[0]);
+                energy_atomic.fetch_add(contrib);
+
+                NESO_PARTICLES_KERNEL_END
+              });
+        })
+        .wait_and_throw();
+
+    this->dh_particle_total_momentum_0.device_to_host();
+    const double tmp_particle_momentum_0 = this->dh_particle_total_momentum_0.h_buffer.ptr[0];
+    double total_particle_momentum_0;
+    MPICHK(MPI_Allreduce(&tmp_particle_momentum_0, &total_particle_momentum_0, 1, MPI_DOUBLE,
+                         MPI_SUM, sycl_target->comm_pair.comm_parent));
+
+    return total_particle_momentum_0;
+  }
+  
+  
+  
+      inline double compute_particle_momentum_1() {
+    auto particle_group = this->particle_sys->particle_group;
+    auto k_ion_p1 = (*particle_group)[Sym<REAL>("ION_MOMENTUM_1")]
+                        ->cell_dat.device_ptr();
+
+    this->dh_particle_total_momentum_1.h_buffer.ptr[0] = 0.0;
+    this->dh_particle_total_momentum_1.host_to_device();
+    auto k_particle_momentum_1 = this->dh_particle_total_momentum_0.d_buffer.ptr;
+
+    const auto pl_iter_range =
+        particle_group->mpi_rank_dat->get_particle_loop_iter_range();
+    const auto pl_stride =
+        particle_group->mpi_rank_dat->get_particle_loop_cell_stride();
+    const auto pl_npart_cell =
+        particle_group->mpi_rank_dat->get_particle_loop_npart_cell();
+
+    sycl_target->queue
+        .submit([&](sycl::handler &cgh) {
+          cgh.parallel_for<>(
+              sycl::range<1>(pl_iter_range), [=](sycl::id<1> idx) {
+                NESO_PARTICLES_KERNEL_START
+                const INT cellx = NESO_PARTICLES_KERNEL_CELL;
+                const INT layerx = NESO_PARTICLES_KERNEL_LAYER;
+
```

Comment:
I wouldn't store momentum when we have mass and velocity already stored on the particles. Unless we need momentum in enough places to amortise the cost of the extra data movement.